Add support for PHP 8.0

Synchro opened the pull request:
This PR adds support for PHP 8.0.

* Tests now run on PHP 7.1, 7,2, 7.3, 7.4 and 8.0 on Ubuntu 18.04 (Bionic)
* Tests run via github actions. Travis-Ci is shutting down soon anyway, and I had trouble with beanstalkd server connections from tests on Travis; it works fine on GitHub, and is faster too.
* Tests updated to use PHPUnit 9.5 syntax. These are dynamically backported to older PHPUnit versions (using yoast/phpunit-polyfills – props to @jrfnl!) so that tests can run on older PHP versions without complicated CI config.
* Some very minor issues fixed, mostly phpdoc, unneeded imports.
* I have disabled code coverage support for now, no uploads to scrutinizer as a result (I've found that scrutinizer isn't that great anyway – better to use psalm/phpstan/phpinsights locally).

SamMousa commented:
Cool, even if local coverage is better, why not have both?

Coverage should work fine on github actions as well...

Synchro commented:
It was mainly that when I tried to set it up by copying the coverage config from travis, it failed and I didn't understand why. Please amend if you have an idea of how to make it work!

Synchro commented:
That scrutinizer failure looks unrelated to the code? The tests themselves are all working.

jrfnl commented:
For what it's worth: I've been using [Coveralls](coveralls.io/) for code coverage myself in a number of projects and have found that to work well in combination with GitHub actions.

Example implementation: https://github.com/PHPCSStandards/PHPCSUtils/blob/develop/.github/workflows/test.yml#L186-L272 (probably more complex implementation than what you'd need)
Example output: https://coveralls.io/github/PHPCSStandards/PHPCSUtils

jrfnl commented on .github/workflows/tests.yaml:
```diff
@@ -0,0 +1,35 @@
+        uses: shivammathur/setup-php@v2
+        with:
+          php-version: ${{ matrix.php }}
+          coverage: xdebug
```
Is xdebug needed when code coverage is not being recorded (for now) ?

Synchro replied:
Probably not, but ideally it would be used to get coverage working! The travis version wasn't using xdebug for coverage, but I'm not sure how to configure it.

SamMousa replied:
pcov is how you get coverage in 2021 not xdebug

jrfnl commented on .github/workflows/tests.yaml:
```diff
@@ -0,0 +1,35 @@
+
+      - name: Run tests
+        run: |
+          ./vendor/bin/phpunit --configuration ./phpunit.xml.dist
```
```suggestion
          ./vendor/bin/phpunit
```
(no need to pass the configuration file if it would be found by default anyway)

jrfnl commented on phpunit.xml.dist:
```diff
@@ -1,25 +1,13 @@
-    </filter>
-
+<phpunit xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" backupGlobals="false" backupStaticAttributes="false" colors="true" convertErrorsToExceptions="true" convertNoticesToExceptions="true" convertWarningsToExceptions="true" processIsolation="false" stopOnFailure="false" bootstrap="tests/bootstrap.php" xsi:noNamespaceSchemaLocation="https://schema.phpunit.de/9.3/phpunit.xsd">
+  <coverage>
```
This is the PHPUnit 9.3 way of Configuring coverage.

It will _only_ work with PHPUnit 9.3+ which will limit the PHP versions from which coverage can be obtained. The "old style" configuration will continue to work until PHPUnit 10 and PHPUnit offers a command line way to "on the fly" update from the "old style" configuration to "new style", so if it is important to be able to retrieve code coverage for older PHP versions, I'd recommend using the "old style" configuration for the time being and if needs be, to update it on the fly to the "new style".

Synchro replied:
I reverted this change.

jrfnl commented on tests/Pheanstalk/BugfixConnectionTest.php:
```diff
@@ -27,7 +27,7 @@ public function testIssue4ReadingOver8192Bytes()
         $pheanstalk->put(str_repeat('.', $length));
         $job = $pheanstalk->peekReady();
-        $this->assertEquals(strlen($job->getData()), $length, 'data length: %s');
+        self::assertEquals(strlen($job->getData()), $length, 'data length: %s');
```
```suggestion
        self::assertSame(strlen($job->getData()), $length, 'data length: %s');
```
(always prefer strict type assertions over loose type)

jrfnl replied:
Note: not sure why the change was made from the `$this->` pattern to the `self::` pattern, but just FYI: both will work with the PHPUnit Polyfills.

Synchro replied:
It was because they are apparently static in newer versions of phpunit, and my IDE was complaining about them.

jrfnl replied:
Ah! Well, those methods were always static, might just be that your IDE got smarter 😂

Synchro commented:
Thanks for the pointers Juliette!

I changed all the `assertEquals` to `assertSame`, then changed back those that didn't pass – these might be a good target for future work as pretty much all of them are int to string comparisons that should be fixed at source.

I've reverted the phpunit config file format change – I only changed it to make phpunit be quiet :)

Synchro commented:
So that doesn't work. It appears to generate the coverage file, download ocular.phar, but (I'm guessing) the upload fails:

```
Generating code coverage report in Clover XML format ... done [172 ms]
16
#!/usr/bin/env php
17
Uploading code coverage for repository "g/Synchro/pheanstalk" and revision "49a1d98b77f940c5a15a07a80fd59ec8f0dfe5bc"... Failed
18
{
19
    "message": "Not found",
20
    "description": "The requested page was not found, or you have no access to view it."
21
}
22
Error: Process completed with exit code 1.
```

On Scrutinizer, [it shows a different problem that seems unrelated to anything in this project](https://scrutinizer-ci.com/g/pheanstalk/pheanstalk/inspections/647842f4-255c-4e2f-acc9-327282ba0756):

```
[Info]: The Log File is not empty, but the Build did not fail. Maybe just warnings got logged. You can review the log in /tmp/php-build.8.0.1.20210130174959.log
[Success]: Built 8.0.1 successfully.
Updating php.ini file...
Updating compiled Apache module...
Updating pecl settings...
config-set (auto_discover, 1, user) failed, channel pecl.php.net
No releases available for package "pecl.php.net/xdebug"
install failed
```

Does it really build PHP from source *every time*???

I don't know what to do to fix this part.

SamMousa commented:
I'll have a look soon, thanks for the PR and the comments!

SamMousa commented on .github/workflows/tests.yaml:
```diff
@@ -0,0 +1,36 @@
+
+      - name: Set up PHP
+        uses: shivammathur/setup-php@v2
+        with:
```
Add pcov here, check the docs of this action

SamMousa commented on .github/workflows/tests.yaml:
```diff
@@ -0,0 +1,36 @@
+          php-version: ${{ matrix.php }}
+
+      - name: Install PHP packages
+        run: composer install --no-interaction --prefer-source
```
Why do we prefer source?

Synchro replied:
It may not be needed – I've occasionally had problems with components not installing properly without that set.

Synchro replied:
I've removed that and it seems happy

SamMousa commented on .github/workflows/tests.yaml:
```diff
@@ -0,0 +1,36 @@
+
+      - name: Run tests
+        run: |
+          phpdbg -qrr vendor/bin/phpunit --coverage-clover clover.xml
```
isn't needed if you use `pcov`

Synchro replied:
I've switched it to use pcov

SamMousa commented on composer.json:
```diff
@@ -1,41 +1,43 @@
+    "ext-mbstring": "*"
+  },
+  "require-dev": {
+    "yoast/phpunit-polyfills": "^0.2.0"
```
shouldn't phpunit itself still be in the require-dev?

Synchro replied:
No, the Yoast polyfill deals with that as it chooses a different version of PHPUnit according to what version of PHP you're running on.

SamMousa commented:
@Synchro  i'm inclined to just up the php required version and the phpunit version in a new major version of this lib.

There is no real active development on it since it is mostly feature complete; trying to stay cross version compatible means more effort even with the polyfill package..

Upping php to 8 would also allow us all the things like typed returns in function declarations from php 7.1+ etc

What are your thoughts?

Synchro commented:
The main part of this PR (apart from the overage thing!) was updating the tests to use phpunit 9.5 syntax. As you say, the code itself is feature complete, but as far as I'm concerned (as a user of your package), the important thing is to get a new version out that's tagged for PHP 8.0, and this PR does that. Once that's done, you're free to drop older PHP versions and update syntax.

The polyfill frees lets you worry less about phpunit versions – for example phpunit 9 has a minimum of PHP 7.3, but as you can see it's working fine with 7.1 builds here. Your minimum PHP version (from your composer file) is already 7.1, so you've already got typed function returns and a bunch of other stuff, and there are some things you could improve without raising the language level. If you were going to drop the polyfills, you'd need to drop support for both PHP 7.1 and 7.2. I'm not sure how much you'd gain by dropping support for older PHP versions – in some cases the differences are very small, so it's really a question of how much active development you want to do.

SamMousa commented:
Why did you close this?

Synchro commented:
Clicked the wrong button!

Synchro commented:
Yay! All green!

alister commented on .gitignore:
```diff
@@ -8,4 +8,5 @@ pheanstalk.phar
 tests/coverage
-.idea
\ No newline at end of file
+.idea
```
Ignoring these editor-sourced directories, and files, is better done on a local user basis. You can setup a `.gitignore` file on your machine, that will also be used in every repository.
The Github help site has an example for [Configuring ignored files for all repositories on your computer](https://docs.github.com/en/github/using-git/ignoring-files#configuring-ignored-files-for-all-repositories-on-your-computer).

I've got a sample global .gitignore file in [one of my own repos](https://github.com/alister/dotfiles/blob/master/gitignore) It is renamed to ~/.gitignore, and setup for use in my ~/.gitconfig file:
```ini
[core]
    excludesfile = /home/username/.gitignore
```

SamMousa replied:
Thanks, I'm not sure I fully agree. Having this here is not big downside, there are only a few common editors and IDEs. (more devil's advocate than a really strong opinion).

Jetbrains themselves go even further and recommend committing part of the folder so   IDE settings like line endings or tabs vs spaces are automatically shared. There is an argument to be made that it provides value to users of that IDE without costing users of other IDEs anything. Of course the other extreme would be to have lots Of IDE specific files or ignores that are essentially unrelated to this project...

I'll give it some thought.
